lib/core-complete.sh

```diff
@@ -1602,6 +1622,8 @@ function ble/complete/source:command {
   [[ $compgen ]] || return 1
   ble/util/assign-array arr 'ble/bin/sort -u <<< "$compgen"' # 1 fork/exec
 
+  ble/complete/source/limit arr
```

Comment:
I think, when it reaches the limit, one can soon exit the function rather than proceeding with the subsequent steps with an empty array. Canceling the processing is logically more natural than proceeding with the empty array.

Reply:
The advantage of emptying the array is that it's not intrusive at all: the check is just a single line that we can insert each time it's needed without worrying too much. IMO early returns bring new code paths (and potentially bugs) as they require handling of exit codes all along the call chain.

Reply:
Thank you! I think it's a matter of taste. but I'd like to maintain the code to be faithful to the idea behind it rather than make the possible future modification easier (such as single-line insertion).

If I correctly understand what is meant by *intrusive*, I think whether it is intrusive or not, in this case, depends on the point of view. Actually, in my mind, the array belongs to the caller function so modifying it inside `ble/complete/source/limit` seems intrusive to me. Also, with this design of `ble/complete/source/limit`, it can only be used when the candidates are stored in arrays. There can be a case that the number of possible candidates is available in advance before the candidates are actually generated and stored in an array, where one wants to test the limit before the actual candidate generation. I thought the helper function should be independent of the implementation detail of the caller function, which is the reason why I let `ble/complete/source/test-limit` in https://github.com/akinomyoga/ble.sh/pull/65#issuecomment-720229429 receive a number (to which the caller can pass the number of array elements or any other values).

As for the early return, it is again a matter of taste, but whether early return should be used or not depends on the case. I can understand that, by refraining from early returns and suppressing the number of code paths, certain types of bugs can be avoided, but there are instead bugs caused by not reducing the problem interest. In fact, in the current PR, `ble/complete/source:command` doesn't correctly return `1` when it surpassed the limit, which is a consequence of handling different cases into a branch. When early returns are not used, one needs to care about all the exceptional cases in modifying any part of the functions.

I'm actually highly skeptical of cyclomatic complexity or alike. It does extract some aspects of the code, but it's not everything that determines the readability or maintainability. Anyway, there are certain cases that I'll absolutely use early returns. In particular, I'll use early returns for error/exceptional cases, and this complete limit check fits this pattern.

Exit code handling is intended design and unrelated to the early returns. Even without early returns, I would request to properly handle exit codes.